collector/processor/batchprocessor/metrics.go

```diff
@@ -0,0 +1,212 @@
+		Name:        processorhelper.BuildCustomMetricName(typeStr, statBatchSendSize.Name()),
+		Measure:     statBatchSendSize,
+		Description: statBatchSendSize.Description(),
+		TagKeys:     processorTagKeys,
+		Aggregation: view.Distribution(10, 25, 50, 75, 100, 250, 500, 750, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 50000, 100000),
+	}
+
+	distributionBatchSendSizeBytesView := &view.View{
+		Name:        processorhelper.BuildCustomMetricName(typeStr, statBatchSendSizeBytes.Name()),
+		Measure:     statBatchSendSizeBytes,
+		Description: statBatchSendSizeBytes.Description(),
+		TagKeys:     processorTagKeys,
+		Aggregation: view.Distribution(10, 25, 50, 75, 100, 250, 500, 750, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 50000,
+			100_000, 200_000, 300_000, 400_000, 500_000, 600_000, 700_000, 800_000, 900_000,
+			1000_000, 2000_000, 3000_000, 4000_000, 5000_000, 6000_000, 7000_000, 8000_000, 9000_000),
+	}
+
+	return []*view.View{
+		countBatchSizeTriggerSendView,
+		countTimeoutTriggerSendView,
+		distributionBatchSendSizeView,
+		distributionBatchSendSizeBytesView,
+	}
+}
+
+type batchProcessorTelemetry struct {
+	level    configtelemetry.Level
+	detailed bool
+	useOtel  bool
+
+	exportCtx context.Context
+
+	processorAttr            []attribute.KeyValue
+	batchSizeTriggerSend     metric.Int64Counter
+	timeoutTriggerSend       metric.Int64Counter
+	batchSendSize            metric.Int64Histogram
+	batchSendSizeBytes       metric.Int64Histogram
+	batchMetadataCardinality metric.Int64ObservableUpDownCounter
+}
+
+func newBatchProcessorTelemetry(set processor.CreateSettings, currentMetadataCardinality func() int, useOtel bool) (*batchProcessorTelemetry, error) {
+	exportCtx, err := tag.New(context.Background(), tag.Insert(processorTagKey, set.ID.String()))
+	if err != nil {
+		return nil, err
+	}
+
+	bpt := &batchProcessorTelemetry{
+		useOtel:       useOtel,
+		processorAttr: []attribute.KeyValue{attribute.String("processor", set.ID.String())},
+		exportCtx:     exportCtx,
+		level:         set.MetricsLevel,
+		detailed:      set.MetricsLevel == configtelemetry.LevelDetailed,
+	}
+
+	if err = bpt.createOtelMetrics(set.MeterProvider, currentMetadataCardinality); err != nil {
+		return nil, err
+	}
+
+	return bpt, nil
+}
+
+func (bpt *batchProcessorTelemetry) createOtelMetrics(mp metric.MeterProvider, currentMetadataCardinality func() int) error {
+	if !bpt.useOtel {
+		return nil
+	}
+
+	var errors, err error
+	meter := mp.Meter(scopeName)
+
+	bpt.batchSizeTriggerSend, err = meter.Int64Counter(
+		processorhelper.BuildCustomMetricName(typeStr, "batch_size_trigger_send"),
+		metric.WithDescription("Number of times the batch was sent due to a size trigger"),
+		metric.WithUnit("1"),
+	)
+	errors = multierr.Append(errors, err)
+
+	bpt.timeoutTriggerSend, err = meter.Int64Counter(
+		processorhelper.BuildCustomMetricName(typeStr, "timeout_trigger_send"),
```

Comment:
I'm not very familiar with the batch processor, so my comment might be off-base. According to the instrumentation, the batch processor has two types of triggers: timeout and max batch size reached. I'm wondering if that's sufficient to control the memory usage of this processor. What happens if the number of distinct metadata value combinations is very high? Could we introduce a third type of trigger that sends the current batches when the total number of entries across all batches reaches a specific threshold in order to keep the overall memory usage under control?

Reply:
Please see #80.

Reply:
Thanks for the review! Yeah this is the first in a series of PRs to enhance the processor. One way of controlling in flight memory usage is that the processor will block requests from being added https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/batch_processor.go#L271 if the processing queue is full. In a later PR memory efficiency will be enhanced by controlling admission to the queue with a semaphore based on the size(in bytes) of the request. Unsure if a new trigger would be necessary, but will keep that in mind as we implement our enhancements and observe memory usage in testing